wp-tevko-responsive-images.php

```diff
@@ -297,31 +297,130 @@ function tevkori_get_src_sizes( $id, $size = 'thumbnail' ) {
  *
- * @see images_send_to_editor
- * @return string HTML for image.
+ * @since 3.0
+ *
+ * @param string $content The raw post content to be filtered.
+ * @return string Converted content with `srcset` and `sizes` added to images.
+ */
+function tevkori_filter_content_images( $content ) {
+	// Only match images in our uploads directory.
+	$uploads_dir = wp_upload_dir();
+	$path_to_upload_dir = $uploads_dir['baseurl'];
+
+	$content = preg_replace_callback( '|<img ([^>]+' . $path_to_upload_dir . '[^>]+)[\s?][\/?]>|i', '_tevkori_filter_content_images_callback', $content );
+
+	return $content;
+}
+add_filter( 'the_content', 'tevkori_filter_content_images', 5, 1 );
+
+/**
+ * Private preg_replace callback used in tevkori_filter_content_images()
+ *
+ * @access private
+ * @since 3.0.0
  */
-function tevkori_extend_image_tag( $html, $id, $caption, $title, $align, $url, $size, $alt ) {
-	add_filter( 'editor_max_image_size', 'tevkori_editor_image_size' );
+function _tevkori_filter_content_images_callback( $image ) {
+	if ( empty( $image ) ) {
+		return false;
+	}
+
+	list( $image_html, $atts ) = $image;
+	$id = $size = false;
+
+	// Bail early if a `srcset` attribute already exists.
+	if ( false !== strpos( $atts, 'srcset=' ) ) {
+		return $image_html;
+	}
 
-	// Get the srcset attribute.
-	$srcset = tevkori_get_srcset_string( $id, $size );
+	// Grab ID and size info from core classes.
+	if ( preg_match( '/wp-image-([0-9]+)/i', $atts, $class_id ) ) {
+		(int) $id = $class_id[1];
+	}
+	if ( preg_match( '/size-([^\s|"]+)/i', $atts, $class_size ) ) {
+		$size = $class_size[1];
+	}
 
-	remove_filter( 'editor_max_image_size', 'tevkori_editor_image_size' );
+	if ( $id && false === $size ) {
+		preg_match( '/width="([0-9]+)"/', $atts, $width );
+		preg_match( '/height="([0-9]+)"/', $atts, $height );
+
+		$size = array(
+			(int) $width[1],
+			(int) $height[1]
+		);
+	}
 
-	if ( $srcset ) {
+	/*
+	 * If attempts to get values for `id` and `size` failed, use the
+	 * src to query for matching values in `_wp_attachment_metadata`
+	 */
+	if ( false === $id || false === $size ) {
+		preg_match( '/src="([^"]+)"/', $atts, $url );
+
+		if ( ! $url[1] ) {
+			return $image_html;
+		}
 
-		// Build the data-sizes attribute if sizes were returned.
-		$sizes = tevkori_get_sizes( $id, $size );
-		$sizes = $sizes ? 'data-sizes="' . $sizes . '"' : '';
+		$image_filename = basename( $url[1] );
+
+		/*
+		 * If we already have an ID, we use it to get the attachment metadata
+		 * using `wp_get_attachment_metadata()`. Otherwise, we'll use the image
+		 * `src` url to query the postmeta table for both the attachement ID and
+		 * metadata, which we'll use later to get the size.
+		 */
+		if ( ! empty( $id ) ) {
+			$meta = wp_get_attachment_metadata( $id );
+		} else {
+			global $wpdb;
+			$meta_object = $wpdb->get_row( $wpdb->prepare(
+				"SELECT `post_id`, `meta_value` FROM $wpdb->postmeta WHERE `meta_key` = '_wp_attachment_metadata' AND `meta_value` LIKE %s",
```

Comment:
We shouldn't be running this on display, since it's bound to cause slowdowns on sites with much media.
If it's going to be run, I'd suggest the information be cached where it can more easily be looked up for the next display.